patches/server/0936-Add-Entity-hidden-by-default-flag.patch

```diff
@@ -0,0 +1,207 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Justin <justin@justinf.dev>
+Date: Wed, 24 Aug 2022 05:17:20 -0700
+Subject: [PATCH] Add Entity hidden by default flag
+
+Adds a boolean to the server Entity controlling if the entity should be hidden by default. 
+The TrackedEntity maintains a set of all players that will be able to see the entity even 
+when it is hidden by default. This set is modified when Player#showEntity and Player#hideEntity 
+are invoked. Changes are made to the way that TrackedEntity updates players when the entity is 
+hidden by default - if a player is not present in the above set, the update is abandoned.
+
+This functionality is expanded when a Player is hidden by default. The player will send out 
+PlayerInfo packets to hide themselves from other Players when hiddenByDefault. There remains a 
+discrepancy when trying to getHiddenPlayers - players that are hidden by default are currently 
+not returned from this method, only those hidden with hideEntity or hidePlayer.
+
+Hiding entities by default also respects precedent of keeping entities hidden when other plugins 
+explicitly hide them from a player. If an entity is hidden by default and by a plugin, the entity 
+must have the plugin that hid them from a player show them in order to be shown to a player again. 
+If no plugin had previously hidden the entity before the entity was hidden by default, then any 
+plugin that attempts to show the entity to a player will succeed in doing so.
+
+diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
+index 77c89376495d90d0e7cbf6cd02c9a1c8d9a4340b..c29c2a57b8bab3dd36af0e4ff21c6f0735c260a9 100644
+--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
++++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
+@@ -2277,6 +2277,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
+         private final int range;
+         SectionPos lastSectionPos;
+         public final Set<ServerPlayerConnection> seenBy = new ReferenceOpenHashSet<>(); // Paper - optimise map impl
++        public final Set<ServerPlayerConnection> showToEvenWhenHiddenByDefault = new ReferenceOpenHashSet<>(); // Paper - Entity hiddenByDefault
+ 
+         public TrackedEntity(Entity entity, int i, int j, boolean flag) {
+             this.serverEntity = new ServerEntity(ChunkMap.this.level, entity, j, flag, this::broadcast, this.seenBy); // CraftBukkit
+@@ -2385,6 +2386,13 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
+                     flag = false;
+                 }
+                 // CraftBukkit end
++
++                // Paper start - Entity hiddenByDefault
```

Comment:
Are we not doubling logic here?
https://github.com/jstnf/Paper/blob/5e3a6cc38bf223eaf36657f7f70b364745ed7f4c/patches/server/0936-Add-Entity-hidden-by-default-flag.patch#L170-L174

already checks if a player is respecting the hidden by default state / show overwrite for hidden by default.

Reply:
Good catch, put this check in before updating canSee.